clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -314,69 +370,118 @@ public void poll(long timeout) throws IOException {
+                if (idleExpiryManager != null)
+                    idleExpiryManager.update(channel.id(), currentTimeNanos);
+
+                try {
+
+                    /* complete any connections that have finished their handshake (either normally or immediately) */
+                    if (isImmediatelyConnected || key.isConnectable()) {
+                        madeProgress = true;
+                        if (channel.finishConnect()) {
+                            this.connected.add(channel.id());
+                            this.sensors.connectionCreated.record();
+                            SocketChannel socketChannel = (SocketChannel) key.channel();
+                            log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
+                                    socketChannel.socket().getReceiveBufferSize(),
+                                    socketChannel.socket().getSendBufferSize(),
+                                    socketChannel.socket().getSoTimeout(),
+                                    channel.id());
+                        } else
+                            continue;
+                    }
 
-                /* complete any connections that have finished their handshake (either normally or immediately) */
-                if (isImmediatelyConnected || key.isConnectable()) {
-                    if (channel.finishConnect()) {
-                        this.connected.add(channel.id());
-                        this.sensors.connectionCreated.record();
-                        SocketChannel socketChannel = (SocketChannel) key.channel();
-                        log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
-                                socketChannel.socket().getReceiveBufferSize(),
-                                socketChannel.socket().getSendBufferSize(),
-                                socketChannel.socket().getSoTimeout(),
-                                channel.id());
-                    } else
-                        continue;
-                }
+                    /* if channel is not ready finish prepare */
+                    if (channel.isConnected() && !channel.ready()) {
+                        madeProgress = true;
+                        channel.prepare();
+                    }
 
-                /* if channel is not ready finish prepare */
-                if (channel.isConnected() && !channel.ready())
-                    channel.prepare();
+                    /* if channel is ready and has bytes to read from socket or buffer, and has no previous
+                       message(s) already staged or otherwise in progress then read from it */
+                    if (channel.ready() && (key.isReadable() || channel.hasBytesBuffered()) && !hasStagedReceive(channel)
+                            && !explicitlyMutedChannels.contains(channel)) {
```

Comment:
a channel can be in explicitelyMuted and in keysWithBytesBuffered at the same time - ssl may try and read several requests at once into stagedReceives. if it reads once request and then has no memory for the next the channel will be in keysWithBytesBuffered. the 1st request out of staged will be moved to completed, causing the channel to be muted when SocketServer picks it up (still on the same thread). 

under this condition channel.hasBytesBuffered() == true and also !hasStagedReceive(channel), causing data to be read for a channel that already has a request in progress